# Сравнение использования приложения Яндекс.Музыка в разных городах

<h1>Содержание<span class="tocSkip"></span></h1>
</a></span><li><span><a  href="#Изучение данных из файла" data-toc-modified-id="Изучение данных из файла"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение данных из файла</a></span><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span></li><li><span><a href="#Исследовательский анализ данных" data-toc-modified-id="Исследовательский анализ данных"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследовательский анализ данных</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы</a></span>

Из предотсавленного файла данных необходимо сделать выводы и сравнить поведение пользователей в Москве и Петербурге.
Необходимо проверьте три гипотезы:
1. Зависит ли активность пользователей в Москве и Петербурге от дня недели.
2. В понедельник утром в Москве преобладают одни жанры, а в Петербурге — другие. Аналогичная ситуация складывается вечером пятницы в разных городах преобладают разные жанры. 
3. Отличаются ли жанровые предпочтения пользователей в Москве и Петербурге.

## Изучение данных из файла

In [1]:
# импорт библиотек

import pandas as pd

In [2]:
# прочитан файл и сохранен в переменной df

df = pd.read_csv('/datasets/yandex_music_project.csv')

In [3]:
# выведены на экран первые 5 строк датафрейма

display(df.head())

userID                        Track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

             City        time        Day  
0  Saint-Petersburg  20:28:33  Wednesday  
1            Moscow  14:07:09     Friday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
4            Moscow  08:34:34     Monday

In [4]:
# получена общая информация о датафрейме 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63848 non-null  object
 2   artist    57876 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Из полученных данных видно, что в таблице 7 столбцов. Тип данных во всех столбцах — `object`.

Согласно документации к данным:
* `userID` — идентификатор пользователя;
* `Track` — название трека;  
* `artist` — имя исполнителя;
* `genre` — название жанра;
* `City` — город пользователя;
* `time` — время начала прослушивания;
* `Day` — день недели.

Также видно, что в названиях столбцов есть нарушения стиля, необходимо привести все значения в нижнему регистру, в наименованиях есть пробелы. Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.


**Выводы**

Представленный датафрейм даёт информацию как о самом треке (название, исполнитель, жанр), так и о пользоватете (город и какую музыку он слушает). Из полученных данных видно, что в названиях столбцов есть нарушения стиля, необходимо привести все значения к нижнему регистру, в наименованиях есть пробелы. Для дальнейшей работы необходимо провести предобработку данных - переименовать все столбцы по правилам хорошего стиля, проверить таблицу на пропуски и дубликаты.

## Предобработка данных

### Стиль заголовков

In [5]:
# выведем на экран информацию о наименовании столбцов

print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


In [6]:
# переименуем столбцы по правилам хорошего стиля

df = df.rename(columns={'  userID': 'user_id', 'Track': 'track','  City  ': 'city', 'Day': 'day'})

In [7]:
# проверяем полученный результат

print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


### Пропуски значений

In [8]:
# проверяем пропуски в датафрейме 

display(df.isna().sum())

user_id       0
track      1231
artist     7203
genre      1198
city          0
time          0
day           0
dtype: int64

Из полученных данных видно, что пропуски встречаются в столбцах track, artist и genre. Все пропущенные значения заменим на 'unknown'. Столбцы track и artist не потребуются для дальнейших расчетов, поэтому замена в этих столбцах точно не повлияет на дальнейшие расчеты. Столбец genre будет использван для дальнейших расчетов, но кол-во пропусков составляет 1,9%, значение крайне мало, поэтому не должно повлиять на качество итоговых результатов.

In [9]:
# заменим пропущенные значения в столбцах track, artist и genre на строку 'unknown'

columns_to_replace = ['track', 'artist', 'genre']
for columns in columns_to_replace:
    df[columns] = df[columns].fillna('unknown')

In [10]:
# проверяем полученный результат

print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


### Дубликаты

In [11]:
# проверяем явные дубликаты

print(df.duplicated().sum())

3826


In [12]:
# удаляем явные дубликаты

df = df.drop_duplicates().reset_index(drop=True)

In [13]:
# проверяем итоговый результат

print(df.duplicated().sum())

0


Также необходимо проверить неявные дубликаты в столбце genre.Так как название одного и того же жанра может быть записано по-разному.

In [14]:
# выводим на экран наименования всех жанров

print(sorted(df['genre'].unique()))

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'alternativepunk', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'author', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'fairyt

Из полученныз данных выявлены неявные дубликаты hip,hop,hip-hop.
Заменим все наименования на единое hiphop

In [15]:
# написана функция для замены наименований жанров

def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genres, correct_genre)

In [16]:
duplicates = ['hip', 'hop', 'hip-hop']
name = 'hiphop'
replace_wrong_genres(duplicates, name)

In [17]:
# проверяем полученные данные

print(sorted(df['genre'].unique()))

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'alternativepunk', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'author', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'fairyt

Предобработка данных проведена, можно переходить к исследовательскому анализу.

## Исследовательский анализ данных

### Сравнение поведения пользователей двух столиц

Сравним пользовательскую активность в разных городах в разные дни недели.

In [18]:
# оценим активность пользователей в каждом городе

activity_number = df.groupby('city')['genre'].count()
print(activity_number)

city
Moscow              42741
Saint-Petersburg    18512
Name: genre, dtype: int64


В Москве прослушиваний больше, чем в Петербурге, полученное значение вполне логично, так как население Москвы больше, чем Петербурга.

In [19]:
# Также посчитаем прослушивания в понедельник, среду и пятницу. 

week_activity = df.groupby('day')['genre'].count()
print(week_activity)

day
Friday       21840
Monday       21354
Wednesday    18059
Name: genre, dtype: int64


Из полученных данных видно, что минимальная активность пользователей наблюдается по средам, максимальная - по пятницам. Но это общая картина, возможно, мы получим другие данные, разделив их по городам.

In [20]:
# функция для прослушивания для заданного дня и города

def number_tracks(day, city):
    track_list = df[df['day'] == day]
    track_list = track_list[track_list['city'] == city]
    track_list_count = track_list['user_id'].count()
    return track_list_count
print(number_tracks('Monday', 'Moscow'))

15740


Вызовем эту функцию шесть раз, меняя значение параметров — так, чтобы получить данные для каждого города в каждый из трёх дней.

In [21]:
number_tracks('Monday','Moscow')

15740

In [22]:
number_tracks('Monday','Saint-Petersburg')

5614

In [23]:
number_tracks('Wednesday','Moscow')

11056

In [24]:
number_tracks('Wednesday','Saint-Petersburg')

7003

In [25]:
number_tracks('Friday','Moscow')

15945

In [26]:
number_tracks('Friday','Saint-Petersburg')

5895

In [27]:
# обобщим данные и получим таблицу

data = [['Moscow', 15740, 11056, 15945],
        ['Saint-Petersburg', 5614, 7003, 5895]]
columns = ['city', 'monday', 'wednesday', 'friday']
general_table = pd.DataFrame(data = data, columns = columns)
display(general_table)

city  monday  wednesday  friday
0            Moscow   15740      11056   15945
1  Saint-Petersburg    5614       7003    5895

Из таблицы видно, что макисмальное количетсво прослушиваний в Москве наблюдается в понедельник и пятницу - в среду наблюдается спад. В Петербурге обратная картина  - пик прослушиваний наблюдается по средам. То есть гипотеза подтверждается, в разные дни недели в разных городах количество прослушиваний отличается.

### Музыка в начале и в конце недели

Проверим вторую гипотезу о том, что утром в понедельник в Москве преобладают одни жанры, а в Петербурге — другие (аналогичная ситуация по пятницам).

In [28]:
moscow_general = df[df['city'] == 'Moscow']# получение таблицы moscow_general из тех строк таблицы df, 
display(moscow_general)# для которых значение в столбце 'city' равно 'Moscow'


user_id                          track                   artist  \
1      55204538    Delayed Because of Accident         Andreas Rönnberg   
4      E2DC1FAE                    Soul People               Space Echo   
6      4CB90AA5                           True             Roman Messer   
7      F03E1C1F               Feeling This Way          Polina Griffith   
8      8FA1D3BE       И вновь продолжается бой                  unknown   
...         ...                            ...                      ...   
61247  83A474E7  I Worship Only What You Bleed  The Black Dahlia Murder   
61248  729CBB09                        My Name                   McLean   
61250  C5E3A0D5                      Jalopiina                  unknown   
61251  321D0506                  Freight Train            Chas McDevitt   
61252  3A64EF84      Tell Me Sweet Little Lies             Monica Lopez   

              genre    city      time        day  
1              rock  Moscow  14:07:09     Friday  
4             dance  Moscow  08:34:34     Monday  
6             dance  Moscow  13:00:07  Wednesday  
7             dance  Moscow  20:47:49  Wednesday  
8            ruspop  Moscow  09:17:40     Friday  
...             ...     ...       ...        ...  
61247  extrememetal  Moscow  21:07:12     Monday  
61248           rnb  Moscow  13:32:28  Wednesday  
61250    industrial  Moscow  20:09:26     Friday  
61251          rock  Moscow  21:43:59     Friday  
61252       country  Moscow  21:59:46     Friday  

[42741 rows x 7 columns]

In [29]:
spb_general = df[df['city'] == 'Saint-Petersburg']
display(spb_general)# получение таблицы spb_general из тех строк таблицы df,
# для которых значение в столбце 'city' равно 'Saint-Petersburg'


user_id                              track              artist  \
0      FFB692EC                  Kamigata To Boots    The Mass Missile   
2        20EC38                  Funiculì funiculà         Mario Lanza   
3      A3DD03C9              Dragons in the Sunset          Fire + Ice   
5      842029A1                          Преданная           IMPERVTOR   
9      E772D5C0                          Pessimist             unknown   
...         ...                                ...                 ...   
61239  D94F810B        Theme from the Walking Dead  Proyecto Halloween   
61240  BC8EC5CF       Red Lips: Gta (Rover Rework)               Rover   
61241  29E04611                       Bre Petrunko       Perunika Trio   
61242  1B91C621             (Hello) Cloud Mountain     sleepmakeswaves   
61249  D08D4A55  Maybe One Day (feat. Black Spade)         Blu & Exile   

            genre              city      time        day  
0            rock  Saint-Petersburg  20:28:33  Wednesday  
2             pop  Saint-Petersburg  20:58:07  Wednesday  
3            folk  Saint-Petersburg  08:37:09     Monday  
5          rusrap  Saint-Petersburg  13:09:41     Friday  
9           dance  Saint-Petersburg  21:20:49  Wednesday  
...           ...               ...       ...        ...  
61239        film  Saint-Petersburg  21:14:40     Monday  
61240  electronic  Saint-Petersburg  21:06:50     Monday  
61241       world  Saint-Petersburg  13:56:00     Monday  
61242    postrock  Saint-Petersburg  09:22:13     Monday  
61249      hiphop  Saint-Petersburg  10:00:00     Monday  

[18512 rows x 7 columns]

Напишем функцию genre_weekday(). Функция должна вернуть информацию о топ-10 жанров тех треков, которые прослушивали в указанный день, в промежутке между двумя отметками времени.

In [30]:
def genre_weekday(table, day, time1, time2):
    genre_df = table[table['day'] == day]
    genre_df = genre_df[genre_df['time'] > time1]
    genre_df = genre_df[genre_df['time'] < time2]
    genre_df_count = genre_df.groupby('genre')['track'].count()
    genre_df_sorted = genre_df_count.sort_values(ascending=False)
    genre_df_sorted = genre_df_sorted.head(10)
    return genre_df_sorted

Сравним результаты функции `genre_weekday()` для Москвы и Санкт-Петербурга в понедельник утром (с 7:00 до 11:00) и в пятницу вечером (с 17:00 до 23:00):

In [31]:
print(genre_weekday(moscow_general,'Monday', '07:00', '11:00'))

genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
Name: track, dtype: int64


In [32]:
print(genre_weekday(spb_general,'Monday', '07:00', '11:00'))

genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
Name: track, dtype: int64


In [33]:
print(genre_weekday(moscow_general,'Friday', '17:00', '23:00'))

genre
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world          208
ruspop         170
alternative    163
classical      163
rusrap         142
Name: track, dtype: int64


In [34]:
print(genre_weekday(spb_general,'Friday', '17:00', '23:00'))

genre
pop            256
electronic     216
rock           216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
Name: track, dtype: int64


При сравнении полученных данных видно, что музыкальные предпочтения в Москве и Петербурге практически не отличаются.Единственное отличие — в московский рейтинг вошёл жанр “world”, а в петербургский — джаз и классика. Для Москва значение unknown попало в топ-10, то есть эти значения всё-таки играют сущетсвунную роль и влияют на расчеты (при презентации данных следует обратить внимание на этот момент и сделать акцент на корректности предоставленных данных).Для понедельника и пятница топ-10 жанров несущетсвенно отличается. То есть разница между Москвой и Петербургом не слишком выражена. В Москве чаще слушают русскую популярную музыку, в Петербурге — джаз.

### Жанровые предпочтения в Москве и Петербурге

Проверим Гипотезу, что у Петербурге чаще слушают рэп, а в Москве — город преобладает поп-музыка.

In [35]:
# cгруппируем таблицу `moscow_general` по жанру и посчитайте прослушивания треков каждого жанра
all_moscow = moscow_general.groupby('genre')['track'].count()
moscow_genres = all_moscow.sort_values(ascending=False)
display(moscow_genres.head(10))

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: track, dtype: int64

In [36]:
# Сгруппируем таблицу spb_general по жанру

all_spb = spb_general.groupby('genre')['track'].count()
spb_general = all_spb.sort_values(ascending=False)
display(spb_general.head(10))

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: track, dtype: int64

**Выводы**

Поп-музыка — самый популярный жанр в Москве,при этом рэп одинаково популярен в Москве и Петербурге.

## Выводы

В результате исследовательского анализа было проведено сравнение использования приложения Яндекс.Музыка в разных городах. Представленный датафрейм даёт информацию как о самом треке (название, исполнитель, жанр), так и о пользоватете (город и какую музыку он слушает). Была проведена предобработка данных, переименованы столбцы (написано "змеиным регистром" и проведены к нижнему регистру без пробелов), данные проверены на наличие пропусков и дубликатов. Выявлено 1198 пропусков в столбце genre, в результате работы обнаружено, что они влияют на дальнейшие выводы, поэтому требуется выяснять, почему эти данные не указаны в датафрейме.
Проверено три гипотезы:

1. Зависит ли активность пользователей в Москве и Петербурге от дня недели

Макисмальное количетсво прослушиваний в Москве наблюдается в понедельник и пятницу - в среду наблюдается спад. В Петербурге обратная картина  - пик прослушиваний наблюдается по средам. То есть гипотеза подтверждается, в разные дни недели в разных городах количество прослушиваний отличается.

2. В понедельник утром в Москве преобладают одни жанры, а в Петербурге — другие. Аналогичная ситуация складывается вечером пятницы в разных городах преобладают разные жанры. 

Музыкальные предпочтения в Москве и Петербурге практически не отличаются.Единственное отличие — в московский рейтинг вошёл жанр “world”, а в петербургский — джаз и классика. Для Москва значение unknown попало в топ-10, то есть эти значения всё-таки играют сущетсвунную роль и влияют на расчеты.Для понедельника и пятница топ-10 жанров несущетсвенно отличается. То есть разница между Москвой и Петербургом не слишком выражена. В Москве чаще слушают русскую популярную музыку, в Петербурге — джаз. Гипотеза не подтвердилась.

3. Отличаются ли жанровые предпочтения пользователей в Москве и Петербурге.

Поп-музыка — самый популярный жанр в Москве,при этом рэп одинаково популярен в Москве и Петербурге.Гипотеза подтвердилась частично.